### PR TITLE
feat(testrunner): v2 testrunner

### PR DESCRIPTION
**Pull request by GustavEikaas**
Trying to give the testrunner a cleaner look, today its very "distracting". I want to use more subtle higlight groups and make it intuitive without explicitly spamming "❌" emojis etc..

Features

- [x] Show solution as root group
- [x] Show csprojects
- [x] Strip namespaces for cleaner look
- [x] Add highlights to sln and project for cleaner look
- [x] Testrunner shows solution and csprojects as default when opening
- [x] Peek stacktrace now opens the file in a floating buffer, indicating the offending line
- [x] Uses F# script to parse xml to json for test result parsing
- [x] Testrun results now gets parsed from xmlfile in TestResults folder rather than from stdout
- [x] ClassData and inline data should be grouped under a new form of group

Issues

- [x] ~Dotnet test -t runs synchronously causing major freeze~
- [x] ~Tests using inline data is not matched correctly when updating status~
- [x] ~Identical namespaces across projects causes issues in status reporting when running solution~
- [x] ~Status aggregation on namespaces/csproject before all tests have finished ruins highlight, given wrong indication~
- [x] ~Tests using inline data is not matched correctly when updating status~
- [x] ~Exclude test projects with no tests~
- [x] ~`Workload updates are available` gets accidently parsed as a test~
- [ ] An error will occur if the json output from the xml file is too large. Apparantly vim.fn.json_decode has a rather low max character limit (this can be solved in the f# script by splitting the output into multiple files)

When running dotnet test on the solution it seems hard to differentiate which tests belongs to which project, solved for now by just running a for loop on the projects in the solution and running in parallel. Uncertain which consequences this will have

## Memory issues with Xml

Initially when testing on a rather small group of tests in a project I ran into memory issues when using vim.fn.json_decode on the converted json. I modified the F# script to remove a lot of unused fields from the outputted json, but I can still imagine that a csproject could potentially have 1000+ tests. Which is guaranteed to crash the vim.fn.json_decode execution. Essentially we are only interested in 3 properties for each unit_test result. `testName`, `outcome` and stacktrace(if failed).

## Option 1

Remove the stacktrace and only output a list of testname and outcome, this would almost guarantee no memory issues for the user. When the peek_stack_trace is called it will either have to re-parse the entire xml or modify the f# script to write all stacktraces to separate files in the initial parsing.

## Option 2

Paginate the json, set an arbitrary size and write a "nextPage" property after x amount of items.
This would result in modifying the F# script to parse the xml, chunk the json and output it into one or more files.

## Option 3

Slice the stacktrace to only get the root "error" and the file name and line number. (temporary fix)

image

Peek stacktrace now opens the offending file with virt text to highlight colors
image

### Duplicate testnames without namespaces within same csproject
I hope this is not a common thing cause i have no idea how to fix it
This seems to be caused by the [TestMethod] attribute. When this attribute is used dotnet test -t will not include the namespace only the name of the test. This is specific to MSTest. 

Support for MSTest might be harder to achieve

### CODE REVIEW

**Comment by franroa** [comment]
This is really cool! I was thinking of proposing you something like this. There is one thing you might want to customize for the v2.
The arguments passed to the dotnet test command are static. I think it would be a good idea to be able to pass the arguments in an array. Then, the user can customize (or the plugin can leverage it for new functionalities).

For example, in the future you might want to add running test with coverage.

In following scenario, I want to check if I have coverage for a given entity (an entity is for example "User", so I want to check if I cover my users actions like add, remove, update, etc...)
```
      {
        "<leader>tce",
        function()
          local function get_entity_name()
            local entity_name = vim.fn.input("Please enter the entity name: ")
            return entity_name
          end
          local task = require("overseer").new_task({
            cmd = "dotnet test "
              .. "--no-build "
              .. "--filter FullyQualifiedName~"
              .. get_entity_name()
              .. "/p:CollectCoverage=true "
              .. "/p:CoverletOutputFormat=lcov "
              .. "/p:CoverletOutput="
              .. LazyVim.root.git()
              .. "/coverage/lcov.info",
            cwd = LazyVim.root.git(),
            components = {
              {
                "on_output_quickfix",
                errorformat = vim.o.grepformat,
                open_height = 8,
                items_only = true,
              },
              -- We don't care to keep this around as long as most tasks
              { "on_complete_dispose", timeout = 30 },
              "default",
            },
          })
          task:start()
        end,
        desc = "Run entity related code with coverage",
        ft = "cs",
      },

```

**Comment by GustavEikaas** [comment]
Yeah it was always my plan to add this, I just wanted to find a way of allowing extra args without colliding with existing args. Maybe its a non issue and up to the user but sending `--filter`  twice will cause an error in the command

**Comment by franroa** [comment]
Yes, I guess sending --filter twice would be a problem.. Another idea would be to create a separate command only for running all tests with coverage. In the end, it will be the same, only takes longer to run.

Other thing I was thinking of was to add a keymap for debugging a test. No idea how difficult it could be to be honest

**Comment by GustavEikaas** [comment]
Two possible issues with debugging a test

- [ ] Resolving the file location of a test and line number
- [ ] Will make nvim-dap sort of a dependency to easy-dotnet.nvim (Which isnt preferred)

I guess there are ways to make it more of an "opt-in" nvim-dap

**Comment by GustavEikaas** [comment]
A little bit of topic but I havent configured dap for debugging tests. Do you have this functionality in your vim config today? Do you use easy-dotnet for path resolutions/helpers?

**Comment by franroa** [comment]
It is an adapter of neotest

**Comment by GustavEikaas** [comment]
Amazing! I didnt know about neotest-dotnet. I have been looking into debugging tests but the issue is with the waiting and attaching part. Cool ill check it out

**Comment by franroa** [comment]
the neotest-dotnet project is really amazing. It has three problems, one of them blocks me:
- This issue: https://github.com/Issafalcon/neotest-dotnet/issues/112, this let crash my laptop
- Filtering in neotest is not working 
- Many times you have the output with tests marked as failed although they passed

**Comment by GustavEikaas** [comment]
I have also faced some issues with marking tests as failed/skipped/passed. The stdout of dotnet cli is hard to parse reliably.

A solid option is to write log files as xml and parse them, the main issue is that the entire test execution will have to finish before results are reflected in the TUI

**Comment by franroa** [comment]
The author of neotest-dotnet also mentions that it is very hard to parse the output. 
In my opinion, if there is a trade-off, I would rather prefer to have the correct output even if this takes a bit longer.

**Comment by franroa** [comment]
So my argument would be the following:
- When I run all tests is to know if I have broken something with my last changes. At that point, I will wait for all test to finish and the go for the failed ones. In that case the time to finish is not that important, but to have the correct output
- When I only run a small group (let say, going by namespace), it doesn't take that long for all tests to finish.

I admit that there is also the chance to run a bigger group of tests where you want to have the output as quick as possible but still in the trade-off I would prefer the correctness over the rapidity

**Comment by GustavEikaas** [comment]
Yeah I have given it a lot of thought the last week but I am at a crossroads where I will either
1. Go all in on parsing stdout and creating a loooot of tests to cover it
2. Admit defeat and go for xml parsing

**Comment by franroa** [comment]
Difficult decision.. The big argument for going all in would be to create a function that other plugins can use and eventually team up. 
Another argument against it is that I would first create a solid testrunner (using the commands on a daily basis) first instead of having too many open parts.

**Comment by franroa** [comment]
what about creating a plugin's test for this case? Just testing the parse function, then we could iterative improve the parser. It would be easier to contribute I think

**Comment by GustavEikaas** [comment]
Yeah, its definitely on the roadmap when the plugin gets a bit more "consolidated". I spend way too much time manually testing stuff. I guess another big pro for using xml is that its more unlikely to have breaking changes. There is no guarantee the stdout will uphold its structure when the dotnet cli evolves. But yeah lua tests will have to be made for all regex parsing of stuff

**Comment by GustavEikaas** [comment]
@franroa Speaking of reliable tools. I may have found a way to guarantee the test results pretty much 100%. Lua or vim itself cant parse xml without using external libs. I wrote a simple f# script that uses newtonsoft to convert the logfile trx from xml to json and then using lua/vim i parse it. This should ensure i dont miss any tests and I should be able to write rock solid code for test results.
Im still testing it but expect something soon

**Comment by franroa** [comment]
That sounds awesome! And good idea. I will test it as soon you deploy it. Testing dotnet with neovim has been my greatest trouble with the ide

**Comment by GustavEikaas** [comment]
@franroa Still a lot of error handling and code refactoring before I can merge but I have tested it in real projects and it seems to work really well. (a bit slow maybe). The smaller the test scope the more noticable it gets. But this is to be expected when calling dotnet fsi i guess🤷‍♂️ 

If you want to test use this commitId: 0dd8af8e016f85412c19bb9feb54a06cce35e759

Otherwise it will probably be merged before the weekend is over

**Comment by franroa** [comment]
Brilliant! I will be working with it today. Thank you so much!

**Comment by GustavEikaas** [comment]
@franroa Do you use MSTest normally?
There seems to be a difference in dotnet test -t when using MSTest. The test names are listed without namespaces

**Comment by franroa** [comment]
Not really, I think only in one project I usually don't need to touch, so for me that is not an issue. Discovering test works perfectly for me.
If you want to add support to MSTest or you think in the future for a v3 of the test runner, you could take a look to this: https://github.com/microsoft/vstest

**Comment by franroa** [comment]
I have tested the commit, and works really well. Found out two issues:
- Integration tests (without namespaces) throw an error
- Some tests stuck on "Running". I don't not yet what they have in common. The majority of them run very well, though.
I general, it works really fast and correctly. Good job! Also the the design looks really nice

**Comment by franroa** [comment]
Another small issue:
When I run a whole folder, and one of the tests stucks on "Running" and the other tests pass, then the folder is marked as passed.
This is probably not an error, since the test stuck on "Running" might have passed as well, and this is only a problem of refreshing the state in the output more than getting the real output

**Comment by GustavEikaas** [comment]
Yeah I have pretty loose validation still but im assuming the xml file is missing entries for all the tests stuck on <Running>. The idea is to throw an error and show that the command failed but I wanted to iron out all the bugs first.

Tests without namespaces seems challenging to handle but I have gotten them to work in my project and the example project you provided. Ill try to add more tests to the test suite and see if I can uncover more bugs.

One known issue is name collision with test without namespaces. But I suspect this is only when using [TestMethod] and MsTest

**Comment by franroa** [comment]
I think there is an issue with the discoverability, which then translates into an issue in the test execution.
Take this example:
discoverability

The test runner is discovering here three tests, because I have three methods. In this case, this is not correct because of the annotation `ClassData` in my test cases:

 ```
    [Theory]
    [ClassData(typeof(InsertCalendarDataTest))]
    public async Task InsertCalendarTest(CalendarInsertRequest insertRequest, Type exceptionType)
    {
    ...
    }

    [Theory]
    [ClassData(typeof(UpdateCalendarDataTest))]
    public async Task UpdateCalendarTest(long calendarId, CalendarUpdateRequest updateRequest, Type exceptionType)
    {
    ...
    }

    [Theory]
    [ClassData(typeof(DeleteCalendarDataTest))]
    public async Task DeleteCalendarTest(long calendarId, Type exceptionType)
    {
     ... 
    }

```

The `ClassData` will then contain many subcases for that `Theory`.

Interesting thing is (and might help with solving the issue), that when I use `InlineData` instead of `ClassData`, it works perfectly:
```
        [Theory]
        [InlineData(false)]
        [InlineData(true)]
        public async void GetCalendar(bool commandHandlerError)
        {
           ...
        }

```

**Comment by GustavEikaas** [comment]
Can you run dotnet test -t on this and check the lines for what output you get. I suspect this is a limitation in dotnet test -t and hard to fix

**Comment by GustavEikaas** [comment]
image
```cs
using System.Collections.Generic;
using Xunit;

namespace test.Class;

// The class that provides the data for the test
public class TestDataGenerator : IEnumerable<object[]>
{
    private readonly List<object[]> _data = new List<object[]>
    {
        new object[] { 1, 2, 3 },
        new object[] { -4, -6, -10 },
        new object[] { 0, 0, 0 }
    };

    public IEnumerator<object[]> GetEnumerator() => _data.GetEnumerator();

    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
}

public class MathTests
{
    // This test method will be run with the data provided by the TestDataGenerator class
    [Theory]
    [ClassData(typeof(TestDataGenerator))]
    public void CanAddNumbers(int value1, int value2, int expectedSum)
    {
        var result = value1 + value2;
        Assert.Equal(expectedSum, result);
    }
}

```

Seems like a bug in parsing test names, The tests are not showing up in my list of tests. I will look at it. Should be a simple fix

**Comment by franroa** [comment]
Well, I guess I could summarize into:
- Parsing names
- Types (namespaces) in the arguments are moved to other groups
- When some tests stuck in a group, the whole group are shown as finished (difficult to say if this is actually an issue, or the stuck tests are just not understood as part of the groups)

**Comment by GustavEikaas** [comment]
I think I have good enough repros to fix this. Thanks for supplying so much feedback, repros! :D

**Comment by franroa** [comment]
You are welcome! thanks for doing such a great plugin!

**Comment by GustavEikaas** [comment]
@franroa The case with ClassData is now fixed. They are grouped under a sort of subgroup

**Comment by franroa** [comment]
Now its better. It still doesnt work though
error

InsertCalendar and UpdateCalendar is still stuck on Running
I think you are almost there!

**Comment by GustavEikaas** [comment]
image
I dont get what you are doing differently than me here to get outliers. Any ideas on how to repro?

**Comment by franroa** [comment]
Maybe adding other classes with classdata tests? If you put more classes that belongs alphabetically under your current one it could appear. I don't know

**Comment by GustavEikaas** [comment]
Found it, its a sorting issue. Will fix

**Comment by GustavEikaas** [comment]
@franroa Works for me now, hopefully that fixed it for you aswell

image

File: lua/easy-dotnet/test-runner/keymaps.lua [hunk 1]
```diff
@@ -1,184 +1,147 @@
...
-      local newSection = line:match(testPattern)
-      if newSection ~= nil then
-        break
-      else
-        table.insert(stackTrace, line)
+local function aggregateStatus(matches)
+  for _, namespace in ipairs(matches) do
+    if (namespace.ref.collapsable == true) then
+      local worstStatus = nil
+      --TODO: check array for worst status
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/keymaps.lua:13]: del comment

File: lua/easy-dotnet/test-runner/render.lua [hunk 2]
```diff
@@ -54,6 +37,50 @@ local function translateIndex(line_num)
...
+  local ns_id = require("easy-dotnet.constants").ns_id
+
+  --Some lines are hidden so tracking the actual line numbers using shadow_index
+  local shadow_index = 0
+  for _, value in ipairs(M.lines) do
+    if value.hidden == false or value.hidden == nil then
+      shadow_index = shadow_index + 1
+      if value.icon == "❌" then
+        vim.api.nvim_buf_add_highlight(M.buf, ns_id, "ErrorMsg", shadow_index - 1, 0, -1)
+      elseif value.icon == "<Running>" then
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/render.lua:51]: Should replace icon with result property

File: lua/easy-dotnet/test-runner/runner.lua [hunk 3]
```diff
@@ -47,29 +59,44 @@ local function expand_test_names_with_flags(test_names)
...
         seen[concatenated] = true
       end
     end
+
+    -- -- Add the full test name with arguments (if any) or just the base name
+    if has_arguments and not seen[full_test_name] then
+      table.insert(expanded,
+        {
+          ns = trim(full_test_name),
+          value = trim(full_test_name):match("([^.]+%b())$"),
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/runner.lua:76]: Trim earlier to avoid trimming multiple places, multiple times

File: lua/easy-dotnet/test-runner/runner.lua [hunk 4]
```diff
@@ -86,6 +113,78 @@ end
...
+      end
+      for _, value in ipairs(lines) do
+        table.insert(win.lines, value)
+      end
+      win.refreshLines()
+    end,
+    ---@param code number
+    on_exit = function(_, code)
+      if code ~= 0 then
+        vim.notify(string.format("Discovering tests for %s failed", project.name))
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/runner.lua:180]: Could add fallback to check if project was built

File: lua/easy-dotnet/test-runner/runner.lua [hunk 5]
```diff
@@ -108,53 +207,57 @@ M.runner = function(options)
...
+        namespace = "",
+        type = "csproject",
+        name = value.name,
+        indent = 2,
+        preIcon = "",
+        hidden = false,
+        collapsable = true,
+        icon = "",
+        expand = {},
+        highlight = "Character"
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/runner.lua:250]: todo define highlight groups

File: lua/easy-dotnet/test-runner/test-parser.lua [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        with
+        | :? System.Exception as ex ->
+            printfn "Error: %s" ex.Message
+            1
+
+main fsi.CommandLineArgs.[1..]
+]]
+
+local ensure_and_get_fsx_path = function()
+  local dir = vim.fs.joinpath(vim.fn.stdpath("data"), "easy-dotnet")
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/test-parser.lua:75]: globalize

File: lua/easy-dotnet/test-runner/runner.lua [hunk 7]
```diff
@@ -47,29 +59,44 @@ local function expand_test_names_with_flags(test_names)
           {
             ns = concatenated,
             value = trim(part),
```
Comment (GustavEikaas) [inline lua/easy-dotnet/test-runner/runner.lua:61]: Is there a way to make a parent property here?

**Comment by franroa** [comment]
Sorting is fixed. Now everything is in the right place. New tests stuck on Running though, still related to the ClassData, not sure what they have in common. Maybe they stuck when a parent stuck? not sure.
Otherwise, its working pretty well

**Comment by GustavEikaas** [comment]
Im at a loss. Matches is an array of the "scope" you are testing. If a value is in the match array the status is set to <Running>. According to the code It will error if it does not find a result for every single match in matches array. The only thing I can think off is if the function throws an exception but surely you would notice that? Can you try to run :messages after one of the tests get stuck?

```lua
local function parse_status(result, test_line)
  --TODO: handle more cases like cancelled etc...
  if result.outcome == "Passed" then
    test_line.icon = resultIcons.passed
  elseif result.outcome == "Failed" then
    test_line.icon = resultIcons.failed
    test_line.expand = vim.split(result.stackTrace, "\n")
  elseif result.outcome == "NotExecuted" then
    test_line.icon = resultIcons.skipped
  else
    test_line.icon = "??"
  end
end


local function parse_log_file(relative_log_file_path, win, matches)
  require("easy-dotnet.test-runner.test-parser").xml_to_json(relative_log_file_path,
    ---@param unit_test_results TestCase[]
    function(unit_test_results)
      for _, match in ipairs(matches) do
        local test_line = match.ref
        if test_line.type == "test" or test_line.type == "subcase" then
          local result = unit_test_results[match.line]
          if result == nil then
            error(string.format("Status of %s was not present in xml file", match.line))
          end
          parse_status(result, test_line)
        end
      end
      aggregateStatus(matches)
      win.refreshLines()
    end)
end


```

**Comment by GustavEikaas** [comment]
At this point I feel like merging this pr and opening a new one where I specifically look into your issue

**Comment by franroa** [comment]
Hi, I think it throwed a notification the first run, then it didn't throw any exception with the runs after that. I could not check too much since I had to leave. Tomorrow I will test again.

In any case, it is really awesome the improvements you have done here. Congratulations!
